edge/pkg/edgehub/task/taskexecutor/node_upgrade.go

```diff
@@ -0,0 +1,159 @@
+	upgradeReq, err = getTaskRequest(taskReq)
+	if err != nil {
+		return
+	}
+
+	if upgradeReq.UpgradeID == "" {
+		err = fmt.Errorf("upgradeID cannot be empty")
+		return
+	}
+	if upgradeReq.Version == version.Get().String() {
+		return fsm.Event{
+			Type:   "Upgrading",
+			Action: api.ActionSuccess,
+		}
+	}
+	err = prepareKeadm(upgradeReq)
+	if err != nil {
+		return
+	}
+	return event
+}
+
+func getTaskRequest(taskReq commontypes.NodeTaskRequest) (*commontypes.NodeUpgradeJobRequest, error) {
+	data, err := json.Marshal(taskReq.Item)
+	if err != nil {
+		return nil, err
+	}
+	var upgradeReq commontypes.NodeUpgradeJobRequest
+	err = json.Unmarshal(data, &upgradeReq)
+	if err != nil {
+		return nil, err
+	}
+	return &upgradeReq, err
+}
+
+func upgrade(taskReq types.NodeTaskRequest) (event fsm.Event) {
+	opts := options.GetEdgeCoreOptions()
+	event = fsm.Event{
+		Type: "Upgrade",
+	}
+	upgradeReq, err := getTaskRequest(taskReq)
+	if err != nil {
+		event.Action = api.ActionFailure
+		event.ErrorMsg = err.Error()
+		return
+	}
+	err = keadmUpgrade(*upgradeReq, opts)
+	if err != nil {
+		event.Action = api.ActionFailure
+		event.ErrorMsg = err.Error()
+	}
+	return
+}
+
+func keadmUpgrade(upgradeReq commontypes.NodeUpgradeJobRequest, opts *options.EdgeCoreOptions) error {
+	klog.Infof("Begin to run upgrade command")
+	upgradeCmd := fmt.Sprintf("keadm upgrade --upgradeID %s --historyID %s --fromVersion %s --toVersion %s --config %s --image %s > /tmp/keadm.log 2>&1",
+		upgradeReq.UpgradeID, upgradeReq.HistoryID, version.Get(), upgradeReq.Version, opts.ConfigFile, upgradeReq.Image)
+
+	// run upgrade cmd to upgrade edge node
+	// use nohup command to start a child progress
+	command := fmt.Sprintf("nohup %s &", upgradeCmd)
+	cmd := exec.Command("bash", "-c", command)
+	s, err := cmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("run upgrade command %s failed: %v, %s", command, err, s)
+	}
+	klog.Infof("!!! Finish upgrade from Version %s to %s ...", version.Get(), upgradeReq.Version)
+	return nil
+}
+
+func prepareKeadm(upgradeReq *commontypes.NodeUpgradeJobRequest) error {
+	config := options.GetEdgeCoreConfig()
+
+	// install the requested installer keadm from docker image
+	klog.Infof("Begin to download version %s keadm", upgradeReq.Version)
+	if config.Modules.Edged.TailoredKubeletConfig.ContainerRuntimeEndpoint == "" {
+		config.Modules.Edged.TailoredKubeletConfig.ContainerRuntimeEndpoint = config.Modules.Edged.RemoteRuntimeEndpoint
+	}
+	container, err := util.NewContainerRuntime(config.Modules.Edged.TailoredKubeletConfig.ContainerRuntimeEndpoint, config.Modules.Edged.TailoredKubeletConfig.CgroupDriver)
+	if err != nil {
+		return fmt.Errorf("failed to new container runtime: %v", err)
+	}
+	image := upgradeReq.Image
+
+	// TODO: do some verification 1.sha256(pass in using CRD) 2.image signature verification
+	// TODO: release verification mechanism
+	err = container.PullImages([]string{image})
```

Comment:
This code is repeated in keadm, is it necessary to abstract it?

Reply:
These two portions pull keadm and edge respectively. It's not the same code.